### DISCUSSION TITLE
Dogfooding and tB...does it apply? how?

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#679] [General]
I remember this term being bandied about in the early .NET development days - often as a criticism of Microsoft for not sufficiently doing so. So, what does it mean?
From: https://www.techopedia.com/definition/30784/dogfooding
What Does Dogfooding Mean?
The term "dogfooding" is an IT slang for the use of one's own products. In some uses, it implies that developers or companies are using their own products to work out bugs, as in beta testing. One benefit of dogfooding is that it shows that a company is confident about its products.

I also recall one critic of programming languages from long ago pointing to languages like BASIC, saying that the only TRUE languages and compilers are those that are able to be written and further developed using that same language and compiler.

So, let me ask this (academic, for now) question: Will twinBasic ever be able to reach that threshold (meaning that the twinBasic compilers is written in twinBASIC code - or at least is able to be)?
...and does "dogfooding" really apply to this effort?

**Comment by bclothier** [comment]
> Will twinBasic ever be able to reach that threshold (meaning that the twinBasic compilers is written in twinBASIC code - or at least is able to be)?

In this specific context, you're actually describing bootstrapping more than dogfooding and yes while it is possible (see: [Rust](https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html)), I think it's probably few years more before that can happen. I _think_ Rust took 10 years before it could be bootstrapped. 

Dogfooding usually refer more to using the same products you produce and is more broad than writing a compiler in its own language.

**Reply by mburns08109** [reply]
I concede the bootstrapping -vs- dogfooding terminology difference (I thought there was another term for that, but couldn't think of it).

:-)

**Reply by bclothier** [reply]
and because I'm a nerd, I remembered that Nim also made the same claims, so looked and one of them did post a walkthrough on [bootstrapping Nim compiler](http://goran.krampe.se/2014/10/15/bootstrapping-nim/). Interestingly, it's not what I picture as bootstrapping because they first generate C source code using Nim to then build the Nim compiler. However, the pipeline seems shorter than Rust's.

**Reply by DavidHollman** [reply]
I recall "self hosting" instead of "bootstrapping". I had heard "bootstrapping" in the context of a new operating system rather than a new language.

**Comment by WaynePhillipsEA** [comment]
Weirdly, I was thinking about this very thing a few days ago.  It's a good question!  

If you were to ask me could the tB compiler/debugger/etc be written in VBx, I'd say... probably yes, though it may not be pretty in places, and it would be difficult to match the performance.

So naturally the answer is the same if we were to rewrite it in tB... though I would certainly hope it would be easier to write, much cleaner code, and performance should be on par once we've got the optimizing compiler in place.

I'm not sure it'll ever happen... but I do hope so as it would be a great challenge!   It may be that we can find a way of doing it gradually?

**Reply by mburns08109** [reply]
That'd be kinda funny - write the tB compiler et. al. in VBx - and the resulting language's code doesn't need VB6RUN.DLL, but the compiler itself does!  *irony! snork!*

**Reply by mburns08109** [reply]
Hey, BTW - do you happen to know the actual date/time you started on this project path (officially-wise, as opposed to in an "I wonder if I could..." sense)?   ...just curious  (and I realize that you're leveraging previous work, so that's a sort of head start - but don't count that officially)

**Comment by wqweto** [comment]
[Introduction &middot; Crafting Interpreters](https://craftinginterpreters.com/introduction.html)

> A compiler reads files in one language, translates them, and outputs files in another language. You can implement a compiler in any language, including the same language it compiles, a process called ***self-hosting***.
>
> You can’t compile your compiler using itself yet, but if you have another compiler for your language written in some other language, you use that one to compile your compiler once. Now you can use the compiled version of your own compiler to compile future versions of itself, and you can discard the original one compiled from the other compiler. This is called ***bootstrapping***, from the image of pulling yourself up by your own bootstraps.

![bootstrap.png](https://craftinginterpreters.com/image/introduction/bootstrap.png)
